src/commands/watch.rs

```diff
@@ -78,182 +73,177 @@ pub struct WatchCommand {
-        let app_manifest = spin_loader::local::raw_manifest_from_file(&app)
-            .await?
-            .into_v1();
+        // Start the the uppificator and buildifier to process notifications from the
+        // `watchexec` tasks.
 
-        // We always want to watch the application manifest
-        let manifest_pattern = WatchPattern::new(
-            app.to_str()
-                .with_context(|| format!("non-unicode manifest path {:?}", app))?
-                .to_owned(),
-            app_dir.as_path(),
-        )?;
-
-        // We want to watch the source code if we aren't skipping the build step
-        let source_patterns = match !self.skip_build {
-            false => vec![],
-            true => app_manifest
-                .components
-                .iter()
-                .filter_map(|c| WatchCommand::create_source_pattern(c, &app_dir))
-                .flatten()
-                .collect::<Result<Vec<WatchPattern>>>()?,
+        let uppificator_handle = tokio::task::spawn(async move { uppificator.run().await });
+        let buildifier_handle = if self.skip_build {
+            tokio::task::spawn(tokio::time::sleep(tokio::time::Duration::MAX))
+        } else {
+            tokio::task::spawn(async move { buildifier.run().await })
         };
 
-        // We want to watch a component's source if it has no build step. If we're skipping
-        // building then we'll watch all component sources.
-        let component_source_patterns = app_manifest
+        // Wait for either the user to stop the watch, or some catastrophe to dump us out.
+
+        _ = ctrlc::set_handler(move || {
+            _ = stop_tx.send(Uuid::new_v4());
+        });
+
+        _ = futures::future::select_all(vec![
+            uppificator_handle,
+            buildifier_handle,
+            artifact_watcher_handle,
+            build_files_watcher_handle,
+        ])
+        .await;
+
+        Ok(())
+    }
+
+    fn spawn_watchexec(
+        &self,
+        manifest_dir: &Path,
+        filterer: impl watchexec::filter::Filterer + 'static,
+        notifier: tokio::sync::watch::Sender<Uuid>,
+        impact_description: &'static str,
+    ) -> anyhow::Result<tokio::task::JoinHandle<()>> {
+        let mut rt = watchexec::config::RuntimeConfig::default();
+        rt.pathset([manifest_dir]);
+        rt.filterer(Arc::new(filterer));
+        rt.action_throttle(Duration::from_millis(self.debounce));
+        rt.on_action(move |action: watchexec::action::Action| {
+            tracing::info!(
+                "spin watch detected changes requiring {impact_description}: {}",
+                paths_of(&action)
+            );
+            _ = notifier.send(Uuid::new_v4());
+            action.outcome(watchexec::action::Outcome::DoNothing);
+            async { Ok::<(), Infallible>(()) }
+        });
+        let watcher = Watchexec::new(watchexec::config::InitConfig::default(), rt)?;
+        let watcher_jh = tokio::task::spawn(async move {
```

Comment:
Hmm, it means `JoinHandle` but `_handle` feels too ambiguous, and `_join_handle` feels quite long.  Changed it to `join_handle` since it's only used on two lines and may the `cargo fmt` gods have mercy on my soul.